### ISSUE TITLE
The data (discrete variables) are expected to be enumerated from 0 when learning BN parameters

### ISSUE THREAD

**Issue by tomkovna** [pgmpy/pgmpy#1258]
### Subject of the issue
It appears that if you do discretisation or binning of the data, the data is expected to be enumerated from 0 and you should have consecutive numbers when fitting the data i.e. for four categires 0,1,2,3 - otherwise for some reason sampling wont work. 

### Your environment
* pgmpy version 0.1.10
* Python version 3.7.7
* Operating System macOS Catalina 10.15.4

### Steps to reproduce
```python 
import numpy as np
import pandas as pd
import networkx as nx

from pgmpy.models import BayesianModel
from pgmpy.estimators import MaximumLikelihoodEstimator
from pgmpy.sampling import BayesianModelSampling


model = BayesianModel([('A', 'B'), 
                       ('C', 'B'),
                       ('D', 'B'), 

                       ('B', 'E'),
                       ('A', 'E'), 
                       ('C', 'E'),

                       ('B', 'F'),
                       ('A', 'F'), 
                       ('C', 'F')
                      ])
A = np.random.randint(1,4,10) # <--- i.e. only changing this line to np.random.randint(0,4,10) 
                                                   # solves the problem
B = np.random.randint(0,2,10)
C = np.random.randint(0,3,10)
D = np.random.randint(0,3,10)
E = np.random.randint(0,2,10)
F = np.random.randint(0,2,10)
data = pd.DataFrame({"A":A,"B":B,"C":C,"D":D,"E":E,"F":F})
model.fit(data, estimator=MaximumLikelihoodEstimator)
nx.draw(model, with_labels=True)
for cpd in model.get_cpds():
    print("CPD of {variable}:".format(variable=cpd.variable))
    print(cpd)
model.check_model()
inference = BayesianModelSampling(model)
samples = inference.likelihood_weighted_sample(size=10, return_type='dataframe')
```

### Expected behaviour
Parameters should be learned with given variable values with no assumption of their enumerating and ordering.

### Actual behaviour
```python
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-37-b3518d6800ce> in <module>
----> 1 samples = inference.likelihood_weighted_sample(size=10, return_type='dataframe')

~/.virtualenvs/test/lib/python3.7/site-packages/pgmpy/sampling/Sampling.py in likelihood_weighted_sample(self, evidence, size, return_type)
    223             if evidence:
    224                 evidence_values = np.vstack([sampled[i] for i in evidence])
--> 225                 cached_values = self.pre_compute_reduce(node)
    226                 weights = list(
    227                     map(lambda t: cached_values[tuple(t)], evidence_values.T)

~/.virtualenvs/test/lib/python3.7/site-packages/pgmpy/sampling/Sampling.py in pre_compute_reduce(self, variable)
    100             states = list(zip(variable_evid, state_combination))
    101             cached_values[state_combination] = variable_cpd.reduce(
--> 102                 states, inplace=False
    103             ).values
    104 

~/.virtualenvs/test/lib/python3.7/site-packages/pgmpy/factors/discrete/CPD.py in reduce(self, values, inplace)
    357         tabular_cpd = self if inplace else self.copy()
    358 
--> 359         super(TabularCPD, tabular_cpd).reduce(values)
    360         tabular_cpd.normalize()
    361 

~/.virtualenvs/test/lib/python3.7/site-packages/pgmpy/factors/discrete/DiscreteFactor.py in reduce(self, values, inplace)
    426         phi = self if inplace else self.copy()
    427         values = [
--> 428             (var, self.get_state_no(var, state_name)) for var, state_name in values
    429         ]
    430 

~/.virtualenvs/test/lib/python3.7/site-packages/pgmpy/factors/discrete/DiscreteFactor.py in <listcomp>(.0)
    426         phi = self if inplace else self.copy()
    427         values = [
--> 428             (var, self.get_state_no(var, state_name)) for var, state_name in values
    429         ]
    430 

~/.virtualenvs/test/lib/python3.7/site-packages/pgmpy/utils/state_name.py in get_state_no(self, var, state_name)
     72         """
     73         if self.state_names:
---> 74             return self.name_to_no[var][state_name]
     75         else:
     76             return state_name

KeyError: 0
```

**Comment by tomkovna** [comment]
Changing the line: 
A = np.random.randint(1,4,10) to:
A = np.random.randint(0,4,10) solves the error, but not the problem :)

**Comment by ankurankan** [comment]
Fixed in #1273
